File: C++/EngineSimulation/EngineSimulation/TestingStand.cpp
#include <thread>
#include "TestingStand.h"

void TestingStand::Watch_T()
{
	double eps = 0.001;
	while (_engine->Is_working())
	{
		if (_engine->Get_T() - _engine->Get_T_overheat() >= eps)
		{
			_engine->Stop();
			break;
		}
	}
}

TestingStand::TestingStand()
{

}

TestingStand::TestingStand(Engine& engine)
{
	_engine = &engine;
}

double TestingStand::StartTesting(double T_environment)
{
	std::thread engine_thread(&Engine::Start, _engine, T_environment);
	std::thread watch_thread(&TestingStand::Watch_T, this);
	
	watch_thread.join();
	engine_thread.join();

	return _engine->Get_working_time();

}

TestingStand::~TestingStand()
{

}

File: C#/EngineSimulation/EngineSimulation/TestingStand.cs
using System.Threading;

namespace EngineSimulation
{
    internal class TestingStand
    {
        Engine _engine;
        double _eps = 0.001;

        public double Eps { get { return _eps; } set { _eps = value; } }
        void WatchT()
        {
            while (true)
            {
                if ( _engine.T - _engine.TOverheat >= _eps)
                {
                    _engine.Stop();
                    return;
                }
                if (_engine.WorkingTime > 0 && !_engine.IsWorking)
                    return;
            }
        }

        public TestingStand(Engine engine)
        {
            _engine = engine;
        }

        public double StartTesting(double TEnviroment)
        {
            var engineThread = new Thread(_engine.Start);
            var watchThread = new Thread(WatchT);
            engineThread.Start(TEnviroment);
            watchThread.Start();
            watchThread.Join();
            engineThread.Join();

            return _engine.WorkingTime;
        }

    }
}
